File: java/L400/LeetCode0474.java
package L400;

public class LeetCode0474 {
    public int findMaxForm(String[] strs, int m, int n) {
        int[][] dp = new int[m + 1][n + 1];
        for (String str : strs) {
            int zero = getZero(str);
            int one = str.length() - zero;
            // 从大到小是为了和原值作比较，从小到大会导致 dp[i - zero][j - one] 变成新值而产生覆盖
            for (int i = m;i >= zero;--i) {
                for (int j = n;j >= one;--j) {
                    dp[i][j] = Math.max(dp[i][j], dp[i - zero][j - one] + 1);
                }
            }
        }
        return dp[m][n];
    }

    private int getZero(String s) {
        int zero = 0;
        for (char c : s.toCharArray()) {
            if (c == '0') {
                ++zero;
            }
        }
        return zero;
    }
}

File: cpp/L400/LeetCode0474.cc
#include <vector>
using std::vector;
#include <string>
using std::string;

class Solution {
public:
    int findMaxForm(vector<string>& strs, int m, int n) {
        vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));
        int ans{0};
        for (string const &str : strs) {
            int m0{0}, n1{0};
            for (char c : str) {
                if (c == '1') {
                    ++n1;
                } else {
                    ++m0;
                }
            }
            for (int i = m;i - m0 >= 0;--i) {
                for (int j = n;j - n1 >= 0;--j) {
                    // make this value max
                    dp[i][j] = std::max(dp[i][j], dp[i - m0][j - n1] + 1);
                    ans = std::max(ans, dp[i][j]);
                }
            }
        }
        return ans;
    }
};
